File: ChatDemo-UI3.0/Code/Controllers/Chats/Views/EMChatToolBar.m
/************************************************************
 *  * Hyphenate
 * __________________
 * Copyright (C) 2016 Hyphenate Inc. All rights reserved.
 *
 * NOTICE: All information contained herein is, and remains
 * the property of Hyphenate Inc.
 */

#import "EMChatToolBar.h"

#import "EMChatRecordView.h"
#import "EMMessageTextView.h"
#import "EMConvertToCommonEmoticonsHelper.h"
#import "EMFaceView.h"

#define kDefaultToolBarHeight 91
#define kDefaultTextViewWidth KScreenWidth - 30.f

@interface EMChatToolBar () <UITextViewDelegate,EMChatRecordViewDelegate,EMFaceDelegate>

@property (strong, nonatomic) UIView *activityButtomView;
@property (nonatomic) BOOL isShowButtomView;

@property (weak, nonatomic) IBOutlet EMMessageTextView *inputTextView;

@property (weak, nonatomic) IBOutlet UIButton *cameraButton;
@property (weak, nonatomic) IBOutlet UIButton *photoButton;
@property (weak, nonatomic) IBOutlet UIButton *emojiButton;
@property (weak, nonatomic) IBOutlet UIButton *recordButton;
@property (weak, nonatomic) IBOutlet UIButton *locationButton;
@property (weak, nonatomic) IBOutlet UIButton *fileButton;
@property (weak, nonatomic) IBOutlet UIButton *sendButton;
@property (weak, nonatomic) IBOutlet UIView *line;

@property (strong, nonatomic) EMChatRecordView *recordView;
@property (strong, nonatomic) EMFaceView *faceView;

@property (strong, nonatomic) NSMutableArray *moreItems;

- (IBAction)cameraAction:(id)sender;
- (IBAction)photoAction:(id)sender;
- (IBAction)locationAction:(id)sender;
- (IBAction)recordAction:(id)sender;
- (IBAction)emojiAction:(id)sender;
- (IBAction)sendAction:(id)sender;

@end

@implementation EMChatToolBar

- (id)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self) {
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(keyboardWillChange:)
                                                     name:UIKeyboardWillChangeFrameNotification
                                                   object:nil];
    }
    return self;
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIKeyboardWillChangeFrameNotification object:nil];
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    
    _inputTextView.layer.borderColor = TextViewBorderColor.CGColor;
    _inputTextView.layer.borderWidth = 0.5;
    
    _cameraButton.left = (KScreenWidth/5 - _cameraButton.width)/2;
    _photoButton.left = (KScreenWidth/5 - _photoButton.width)/2 + KScreenWidth/5 * 1;
    _emojiButton.left = (KScreenWidth/5 - _emojiButton.width)/2 + KScreenWidth/5 * 2;
    _recordButton.left = (KScreenWidth/5 - _recordButton.width)/2 + KScreenWidth/5 * 3;
    _locationButton.left = (KScreenWidth/5 - _locationButton.width)/2 + KScreenWidth/5 * 4;
    
    _inputTextView.placeHolder = NSLocalizedString(@"chat.placeHolder", @"Send Message");
    _inputTextView.placeHolderTextColor = CoolGrayColor;
    _line.width = KScreenWidth;
    
    _sendButton.left = KScreenWidth - _sendButton.width - 15.f;
}

- (void)drawRect:(CGRect)rect
{
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextSetFillColorWithColor(context, [UIColor clearColor].CGColor);
    CGContextFillRect(context, rect);

    _inputTextView.width = kDefaultTextViewWidth;
}
#pragma mark - getter

- (EMChatRecordView*)recordView
{
    if (_recordView == nil) {
        _recordView = (EMChatRecordView*)[[[NSBundle mainBundle]loadNibNamed:@"EMChatRecordView" owner:nil options:nil] firstObject];
        _recordView.delegate = self;
    }
    return _recordView;
}

- (EMFaceView*)faceView
{
    if (_faceView == nil) {
        _faceView = [[EMFaceView alloc] initWithFrame:CGRectMake(0, 0, self.frame.size.width, 180)];
        [_faceView setDelegate:self];
        _faceView.backgroundColor = [UIColor colorWithRed:240 / 255.0 green:242 / 255.0 blue:247 / 255.0 alpha:1.0];
        _faceView.autoresizingMask = UIViewAutoresizingFlexibleTopMargin;
    }
    return _faceView;
}

- (NSMutableArray*)moreItems
{
    if (_moreItems == nil) {
        _moreItems = [NSMutableArray arrayWithArray:@[self.cameraButton,self.photoButton,self.emojiButton,self.recordButton,self.locationButton,self.fileButton]];
    }
    return _moreItems;
}

#pragma mark - UITextViewDelegate

- (BOOL)textView:(UITextView *)textView shouldChangeTextInRange:(NSRange)range replacementText:(NSString *)text
{
    return YES;
}

- (BOOL)textViewShouldBeginEditing:(UITextView *)textView
{
    for (UIButton *btn in self.moreItems) {
        btn.selected = NO;
    }
    return YES;
}

- (void)textViewDidBeginEditing:(UITextView *)textView
{
    _sendButton.hidden = NO;
    [UIView animateWithDuration:0.25 animations:^{
        [_inputTextView setNeedsDisplay];
        _inputTextView.width = kDefaultTextViewWidth - _sendButton.width - 15.f;
    }];
}

- (void)textViewDidEndEditing:(UITextView *)textView
{
    _sendButton.hidden = YES;
    [UIView animateWithDuration:0.25 animations:^{
        [_inputTextView setNeedsDisplay];
        _inputTextView.width = kDefaultTextViewWidth;
    }];
}

- (void)textViewDidChange:(UITextView *)textView
{
    [textView setNeedsDisplay];
}

#pragma mark - EMFaceDelegate

- (void)selectedFacialView:(NSString *)str isDelete:(BOOL)isDelete
{
    NSString *chatText = self.inputTextView.text;
    
    if (!isDelete && str.length > 0) {
        self.inputTextView.text = [NSString stringWithFormat:@"%@%@",chatText,str];
    } else {
        if (chatText.length >= 2) {
            NSString *subStr = [chatText substringFromIndex:chatText.length-2];
            if ([self.faceView stringIsFace:subStr]) {
                self.inputTextView.text = [chatText substringToIndex:chatText.length-2];
                [self textViewDidChange:self.inputTextView];
                return;
            }
        }
        if (chatText.length > 0) {
            self.inputTextView.text = [chatText substringToIndex:chatText.length-1];
        }
    }
    
    [self textViewDidChange:self.inputTextView];
}

- (NSMutableAttributedString*)backspaceText:(NSMutableAttributedString*) attr length:(NSInteger)length
{
    NSRange range = [self.inputTextView selectedRange];
    if (range.location == 0) {
        return attr;
    }
    [attr deleteCharactersInRange:NSMakeRange(range.location - length, length)];
    return attr;
}

- (void)sendFace
{
    NSString *chatText = self.inputTextView.text;
    if (chatText.length > 0) {
        if ([self.delegate respondsToSelector:@selector(didSendText:)]) {
            if (![_inputTextView.text isEqualToString:@""]) {
                NSMutableString *attStr = [[NSMutableString alloc] initWithString:self.inputTextView.attributedText.string];
                [self.delegate didSendText:attStr];
                self.inputTextView.text = @"";
            }
        }
    }
}

#pragma mark - EMChatRecordViewDelegate

- (void)didFinishRecord:(NSString *)recordPath duration:(NSInteger)duration
{
    if (self.delegate && [self.delegate respondsToSelector:@selector(didSendAudio:duration:)]) {
        [self.delegate didSendAudio:recordPath duration:duration];
    }
}

#pragma mark - UIKeyboardNotification
- (void)keyboardWillChange:(NSNotification *)aNoti {
    NSDictionary *userInfo = aNoti.userInfo;
    CGRect endFrame = [userInfo[UIKeyboardFrameEndUserInfoKey] CGRectValue];
    CGRect beginFrame = [userInfo[UIKeyboardFrameBeginUserInfoKey] CGRectValue];
    CGFloat duration = [userInfo[UIKeyboardAnimationDurationUserInfoKey] doubleValue];
    UIViewAnimationCurve curve = [userInfo[UIKeyboardAnimationCurveUserInfoKey] integerValue];
    if (CGRectEqualToRect(beginFrame, endFrame)) {
        return;
    }
    void(^animations)() = ^{
        [self _willShowKeyboardFromFrame:beginFrame toFrame:endFrame];
    };
    [UIView animateWithDuration:duration delay:0.0f
                        options:(curve << 16 | UIViewAnimationOptionBeginFromCurrentState)
                     animations:animations completion:nil];
}

#pragma mark - action

- (IBAction)cameraAction:(id)sender
{
    if (self.delegate && [self.delegate respondsToSelector:@selector(didTakePhotos)]) {
        [self.delegate didTakePhotos];
    }
}

- (IBAction)photoAction:(id)sender
{
    if (self.delegate && [self.delegate respondsToSelector:@selector(didSelectPhotos)]) {
        [self.delegate didSelectPhotos];
    }
}

- (IBAction)emojiAction:(id)sender
{
    UIButton *button = (UIButton*)sender;
    button.selected = !button.selected;
    for (UIButton *btn in self.moreItems) {
        if (button != btn) {
            btn.selected = NO;
        }
    }
    if (button.selected) {
        [self.inputTextView resignFirstResponder];
        [self _willShowBottomView:self.faceView];
    } else {
        [self _willShowBottomView:nil];
    }
}

- (IBAction)locationAction:(id)sender
{
    if (self.delegate && [self.delegate respondsToSelector:@selector(didSelectLocation)]) {
        [self.delegate didSelectLocation];
    }
}

- (IBAction)recordAction:(id)sender
{
    UIButton *button = (UIButton*)sender;
    button.selected = !button.selected;
    for (UIButton *btn in self.moreItems) {
        if (button != btn) {
            btn.selected = NO;
        }
    }
    
    if (button.selected) {
        [self.recordView setHeight:187.f];
        [self.inputTextView resignFirstResponder];
        [self _willShowBottomView:self.recordView];
    } else {
        [self _willShowBottomView:nil];
    }
}

- (IBAction)sendAction:(id)sender
{
    if (_inputTextView.text.length > 0) {
        if (self.delegate && [self.delegate respondsToSelector:@selector(didSendText:)]) {
            [self.delegate didSendText:[EMConvertToCommonEmoticonsHelper convertToCommonEmoticons:_inputTextView.text]];
        }
        _inputTextView.text = @"";
    }
}

#pragma mark - private

- (void)_willShowKeyboardFromFrame:(CGRect)beginFrame toFrame:(CGRect)toFrame
{
    NSLog(@"begin %@ -- end %@", NSStringFromCGRect(beginFrame), NSStringFromCGRect(toFrame));
    if (beginFrame.origin.y == KScreenHeight) {
        [self _willShowBottomHeight:toFrame.size.height];
        if (self.activityButtomView) {
            [self.activityButtomView removeFromSuperview];
        }
        self.activityButtomView = nil;
    } else if (toFrame.origin.y == KScreenHeight) {
        [self _willShowBottomHeight:0];
    } else{
        [self _willShowBottomHeight:toFrame.size.height];
    }
}

- (void)_willShowBottomView:(UIView *)bottomView
{
    if (![self.activityButtomView isEqual:bottomView]) {
        CGFloat bottomHeight = bottomView ? bottomView.height : 0;
        [self _willShowBottomHeight:bottomHeight];
        if (bottomView) {
            bottomView.top = kDefaultToolBarHeight;
            [self addSubview:bottomView];
        }
        if (self.activityButtomView) {
            [self.activityButtomView removeFromSuperview];
        }
        self.activityButtomView = bottomView;
    }
}

- (void)_willShowBottomHeight:(CGFloat)bottomHeight
{
    [self _willShowViewFromHeight:bottomHeight];
}

- (void)_willShowViewFromHeight:(CGFloat)toHeight
{
    if (self.delegate && [self.delegate respondsToSelector:@selector(chatToolBarDidChangeFrameToHeight:)]) {
        [self.delegate chatToolBarDidChangeFrameToHeight:toHeight];
    }
}

@end

File: swift-demo/Hyphenate-Demo-Swift/Code/Controllers/Chats/Views/EMChatToolBar.swift
//
//  EMChatToolBar.swift
//  Hyphenate-Demo-Swift
//
//  Created by dujiepeng on 2017/6/19.
//  Copyright © 2017 dujiepeng. All rights reserved.
//

import UIKit

@objc protocol EMChatToolBarDelegate {
    @objc optional
    func didSendText(text:String)
    func didSendAudio(recordPath: String, duration:Int)
    func didTakePhotos()
    func didSelectPhotos()
    func didSelectLocation()
    func didUpdateInputTextInfo(inputInfo: String)
    
    @objc
    func chatToolBarDidChangeFrame(toHeight height: CGFloat)
}

let kDefaultToolBarHeight = 83
let kDefaultTextViewWidth = kScreenWidth - 30

class EMChatToolBar: UIView , UITextViewDelegate, EMChatRecordViewDelegate, EMFaceDelegate{
    
    weak var delegate: EMChatToolBarDelegate?
    
    private var activityButtonView:UIView?
    private var isShowButtomView = false
    
    @IBOutlet weak var inputTextView: EMMessageTextView!
    
    @IBOutlet weak var sendButton: UIButton!
    @IBOutlet weak var cameraButton: UIButton!
    @IBOutlet weak var photoButton: UIButton!
    @IBOutlet weak var emojiButton: UIButton!
    @IBOutlet weak var recordButton: UIButton!
    @IBOutlet weak var locationButton: UIButton!
    @IBOutlet weak var fileButton: UIButton!
    @IBOutlet weak var line: UIView!
    
    lazy var recordView: EMChatRecordView = {()-> EMChatRecordView in
        let chatRecordView = Bundle.main.loadNibNamed("EMChatRecordView", owner: nil, options: nil)?.first as! EMChatRecordView
        chatRecordView.delegate = self
        return chatRecordView
    }()

    lazy var faceView: EMFaceView = {() -> EMFaceView in
        let _faceView = EMFaceView.init(frame: CGRect(x: 0, y: 0, width: self.width(), height: 180))
        _faceView.delegate = self
        _faceView.backgroundColor = UIColor.init(colorLiteralRed: 240 / 255, green: 242 / 255, blue: 247 / 255, alpha: 1)
        
        return _faceView
    }()
    
    lazy var moreItems: Array<UIButton> = {()-> Array<UIButton> in
        let ary = [self.cameraButton, self.photoButton, self.emojiButton, self.recordButton, self.locationButton, self.fileButton]
        return ary as! Array<UIButton>
    }()
    
    class func instance() -> EMChatToolBar {
        let toolBar = Bundle.main.loadNibNamed("EMChatToolBar", owner: nil, options: nil)?.first as? EMChatToolBar
        return toolBar!
    }
    
    public func setupInput(textInfo: String?) {
        inputTextView.set(text: textInfo)
    }
    
    public func fetchInputTextInfo() -> String? {
        return inputTextView.text
    }

    override func awakeFromNib() {
        super.awakeFromNib()
        inputTextView.delegate = self
         NotificationCenter.default.addObserver(self, selector: #selector(chatKeyboardWillChangeFrame), name: NSNotification.Name.UIKeyboardWillChangeFrame, object: nil)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self, name: NSNotification.Name.UIKeyboardWillChangeFrame, object: nil)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        inputTextView.layer.borderColor = TextViewBorderColor.cgColor
        inputTextView.layer.borderWidth = 0.5
        
        cameraButton.left(left: (kScreenWidth/5 - cameraButton.width()) / 2)
        photoButton.left(left: (kScreenWidth/5 - photoButton.width()) / 2 + kScreenWidth / 5 * 1)
        emojiButton.left(left: (kScreenWidth/5 - emojiButton.width()) / 2 + kScreenWidth / 5 * 2)
        recordButton.left(left: (kScreenWidth/5 - recordButton.width()) / 2 + kScreenWidth / 5 * 3)
        locationButton.left(left: (kScreenWidth/5 - locationButton.width()) / 2 + kScreenWidth / 5 * 4)
        
        inputTextView.set(placeHolder: "Send Message")
        inputTextView.set(plcaeHolderTextColor: CoolGrayColor)
        line.width(width: kScreenWidth)

        sendButton.left(left: kScreenWidth - sendButton.width() - 15)
    }
    
    override func draw(_ rect: CGRect) {
        let ctx = UIGraphicsGetCurrentContext()
        ctx?.setFillColor(UIColor.clear.cgColor)
        ctx?.fill(rect)
        
        inputTextView.width(width: kDefaultTextViewWidth)
    }
    
    // MARK: - getter
    
    
    // MARK: - UIKeyboardNotification
    func chatKeyboardWillChangeFrame(noti: Notification) {
        let userInfo = (noti.userInfo as! Dictionary<String, NSValue>)
        let endFrame = userInfo[UIKeyboardFrameEndUserInfoKey]!.cgRectValue
        let beginFrame = userInfo[UIKeyboardFrameBeginUserInfoKey]!.cgRectValue
        let duration = (userInfo[UIKeyboardAnimationDurationUserInfoKey]! as! NSNumber).doubleValue
        weak var weakSelf = self
        UIView.animate(withDuration: duration, delay: 0, options: [.curveEaseInOut, .beginFromCurrentState], animations: { 
            weakSelf?._willShowKeyboard(fromFrame: beginFrame, endFrame: endFrame)
        }, completion: nil)
        
    }
    
    // MARK: - Actions
    
    @IBAction func sendAction(_ sender: UIButton) {
        if inputTextView.text.characters.count > 0 {
            if delegate != nil {
                delegate?.didSendText!(text: inputTextView.text)
            }
            inputTextView.text = "";
        }
    }
    
    @IBAction func cameraAction(_ sender: UIButton) {
        if delegate != nil {
            delegate?.didTakePhotos()
        }
    }
    
    @IBAction func photoAction(_ sender: UIButton) {
        if delegate != nil {
            delegate?.didSelectPhotos()
        }
    }
    
    @IBAction func emojiAction(_ sender: UIButton) {
        sender.isSelected = !sender.isSelected
        for  btn in moreItems {
            if btn != sender {
                btn.isSelected = false
            }
        }
        
        if sender.isSelected {
            inputTextView.resignFirstResponder()
            _willShow(bottomView: faceView)
        } else {
            _willShow(bottomView: nil)
        }
    }
    
    @IBAction func recordAction(_ sender: UIButton) {
        sender.isSelected = !sender.isSelected
        for  btn in moreItems {
            if btn != sender {
                btn.isSelected = false
            }
        }
        
        if sender.isSelected {
            recordView.height(height: 187)
            recordView.width(width: kScreenWidth)
            inputTextView.resignFirstResponder()
            _willShow(bottomView: recordView)
        } else {
            _willShow(bottomView: nil)
        }
    }
    
    @IBAction func locationAction(_ sender: UIButton) {
        if delegate != nil {
            delegate?.didSelectLocation()
        }
    }
    
    @IBAction func fileAction(_ sender: UIButton) {
        if delegate != nil {
            delegate?.didTakePhotos()
        }
    }
    
    
    // MARK: - UITextViewDelegate
    func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
        return true
    }
    
    func textViewShouldBeginEditing(_ textView: UITextView) -> Bool {
        for btn in moreItems {
            btn.isSelected = false
        }
        return true
    }
    
    func textViewDidBeginEditing(_ textView: UITextView) {
        sendButton.isHidden = false
        weak var weakSelf = self
        UIView.animate(withDuration: 0.25) { 
            weakSelf?.inputTextView.setNeedsDisplay()
            weakSelf?.inputTextView.width(width: kDefaultTextViewWidth - self.sendButton.width() - 15)
        }
    }
    
    func textViewDidEndEditing(_ textView: UITextView) {
        sendButton.isHidden = true
        weak var weakSelf = self
        UIView.animate(withDuration: 0.25) {
            weakSelf?.inputTextView.setNeedsDisplay()
            weakSelf?.inputTextView.width(width: kDefaultTextViewWidth)
        }
    }
    
    func textViewDidChange(_ textView: UITextView) {
        textView.setNeedsDisplay()
    }
    
    // MARK: - EMFaceDelegate
    func selected(facialStr: String?, isDelete: Bool) {
        let chatText = inputTextView.text
        if !isDelete && (facialStr?.characters.count)! > 0{
            inputTextView.set(text: chatText! + facialStr!)
        }else {
            let subStr = chatText! as NSString
            if subStr.length >= 2 {
                let str = subStr.substring(from: subStr.length - 2)
                if faceView.stringIsFace(str: str as String) {
                    inputTextView.set(text: subStr.substring(to: subStr.length - 2) as String)
                    return
                }
            }
            
            if subStr.length > 0  {
                inputTextView.set(text: subStr.substring(to: subStr.length - 1) as String)
            }
        }
    }
    
    func backspace(text attr: NSMutableAttributedString, length: Int) -> NSMutableAttributedString{
        let range = inputTextView.selectedRange
        if range.location == 0 {
            return attr
        }
        
        attr.deleteCharacters(in: NSRange.init(location: range.location - length, length: length))
        return attr
    }
    
    func sendFace() {
        if delegate != nil {
            if inputTextView.text.characters.count != 0 {
                let attStr = NSMutableString.init(string: inputTextView.attributedText.string)
                delegate?.didSendText!(text: attStr as String)
                inputTextView.set(text: "")
            }
        }
    }
    
    func sendFace(withEmoji emoji: String) {
        
    }
    
    // MARK: - EMChatRecordViewDelegate
    func didFinish(recordPatch: String, duration: Int) {
        if delegate != nil {
            delegate?.didSendAudio(recordPath: recordPatch, duration: duration)
        }
    }
    
    // MARK: - Private
    private func _willShowKeyboard(fromFrame beginFrame: CGRect, endFrame toFrame: CGRect) {
        if beginFrame.origin.y == kScreenHeight {
            _willShow(bottomHeight: toFrame.size.height)
            if activityButtonView != nil {
                activityButtonView!.removeFromSuperview()
            }
            activityButtonView = nil
        }
        else if toFrame.origin.y == kScreenHeight {
            _willShow(bottomHeight: 0)
        }
        else {
            _willShow(bottomHeight: toFrame.size.height)
        }
    }
    
    private func _willShow(bottomView: UIView?) {
        if activityButtonView != bottomView {
            let bottomHeight = bottomView != nil ? bottomView?.height() : 0
            _willShow(bottomHeight: bottomHeight!)
            
            if bottomView != nil {
                bottomView?.top(top: CGFloat(kDefaultToolBarHeight))
                addSubview(bottomView!)
            }
            
            if activityButtonView != nil {
                activityButtonView!.removeFromSuperview()
            }
            
            activityButtonView = bottomView
        }
    }
    
    private func _willShow(bottomHeight: CGFloat) {
        let fromFrame = frame
        let toHeight = CGFloat(kDefaultToolBarHeight) + bottomHeight
        let toFrame = CGRect(x: fromFrame.origin.x, y: fromFrame.origin.y + (fromFrame.size.height - toHeight), width: fromFrame.size.width, height: toHeight)
        
        if bottomHeight == 0 && frame.size.height == CGFloat(kDefaultToolBarHeight) {
            return
        }
        
        if bottomHeight == 0 {
            isShowButtomView = false
        }else{
            isShowButtomView = true
        }
        
        weak var weakSelf = self
        UIView.animate(withDuration: 0.25) { 
            weakSelf?.frame = toFrame
        }
        
        _willShowView(toHeight: toHeight)
    }
    
    private func _willShowView(toHeight: CGFloat) {
        if delegate != nil  {
            delegate!.chatToolBarDidChangeFrame(toHeight: toHeight)
        }
    }
}
